lib/summary.js

```diff
@@ -0,0 +1,80 @@
+/*
+ * == BSD2 LICENSE ==
+ * Copyright (c) 2014, Tidepool Project
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the associated License, which is identical to the BSD 2-Clause
+ * License as published by the Open Source Initiative at opensource.org.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the License for more details.
+ *
+ * You should have received a copy of the License along with this program; if
+ * not, you can obtain one from Tidepool Project at tidepool.org.
+ * == BSD2 LICENSE ==
+ */
+
+'use strict';
+
+const datumTypeToSummaryType = {
+  'cbg': 'cgm',
+  'smbg': 'bgm',
+};
+
+// Upload Batch Size
+const batchSize = 1000;
+
+// Make sure we can distinguish jellyfish vs platform uploads
+const outdatedReasonDataAdded = 'LEGACY_DATA_ADDED';
+const outdatedReasonUploadCompleted = 'LEGACY_UPLOAD_COMPLETED';
+
+// If uploading an incomplete batch consider the upload session as completed for the purposes of summary recalculation
+function getOutdatedReason(data) {
+  data = Array.isArray(data) ? data : [data];
+  return data.length % batchSize === 0 ? outdatedReasonDataAdded : outdatedReasonUploadCompleted;
```

Comment:
[Jamie and I were discussing the outstanding Jellyfish PRs since he is about to create PRs for all of the Jellyfish migration work, and I happened to read this PR description and see this code...]

What happens if the total number of data uploaded by uploader just happens to be evenly divisible by the batch size? For example, the batch size is 100 and the total number of data uploaded by uploader is 500. Will this miss the "complete" message?

Reply:
Yes, in this case we will not set the outdated reason to `LEGACY_UPLOAD_COMPLETED`, but we will set it to `LEGACY_DATA_ADDED` and queue the summary for recalculation after 90 seconds. If we receive a subsequent batch that's not divisible by 1000, we will then set outdated reason to `LEGACY_UPLOAD_COMPLETED` and queue the summary for recalculation immediately.

The summary is only queued for recalculation after all of the data in the batch has been ingested and is set at 90s so it allows for the next batch to be fully processed before the summary is recalculated. The request timeout for a single batch is 60 seconds (set by linkerd). We may need to verify that this is correctly enforced, because there are various timeouts enforced by the ELB, Gloo and Linkerd.